crates/ruff_python_semantic/src/module.rs

```diff
@@ -0,0 +1,284 @@
+use std::fmt::Formatter;
+use std::ops::Deref;
+use std::sync::Arc;
+
+use ruff_db::file_system::FileSystemPath;
+use ruff_db::vfs::{VfsFile, VfsPath};
+
+use crate::Db;
+
+pub mod resolver;
+
+/// A module name, e.g. `foo.bar`.
+///
+/// Always normalized to the absolute form (never a relative module name, i.e., never `.foo`).
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct ModuleName(smol_str::SmolStr);
+
+impl ModuleName {
+    #[inline]
+    pub fn new(name: &str) -> Self {
+        assert!(!name.is_empty());
+        assert!(!name.starts_with('.'), "module name must be absolute");
+        assert!(!name.ends_with('.'), "module cannot end with a '.'");
+
+        Self(smol_str::SmolStr::new(name))
+    }
```

Comment:
(Doesn't have to be dealt with here.) Another option might be to return a `Result` from this `new` method (similar to `regex::Regex::new()`). Then we could also add other assertions here that would also make sense, such as the fact that all components should be valid Python identifiers

Reply:
What I worry is that it makes the API fairly annoying to use. I need to check where we call `new` to see if returning `None` in these cases (or an error) would make the API cumbersome to use.

Reply:
I changed `new` to return an `Option` and added a new `new_static` method for when working with static module names ("avoids allocations all together)